### PR TITLE
Add swarm_test to e2e test suite

### PR DESCRIPTION
**Pull request by shaominchen**
Add swarm_test to e2e test suite: fixes issue #1249.

Passed all e2e tests locally.

### CODE REVIEW

**Comment by shuklanirdesh82** [comment]
@shaominchen can you please increase default timeout for running our e2e test to 30 mins `-timeout 30m`?

File: tests/e2e/swarm_test.go [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+// 12. Verify the volume is in attached status
+// 13. Delete the volume - expect fail
+// 14. Remove the service
+// 15. Verify the service is gone
+// 16. Verify the volume is in detached status
+func (s *SwarmTestSuite) TestDockerSwarm(c *C) {
+	log.Printf("START: swarm_test.TestDockerSwarm")
+
+	fullVolumeName := verification.GetFullVolumeName(s.master, s.volumeName)
+	opts := "--replicas 1 --mount type=volume,source=" + fullVolumeName + ",target=/vol,volume-driver=vsphere busybox tail -f /dev/null"
```
Comment (shuklanirdesh82) [inline tests/e2e/swarm_test.go]: minor: better to reuse const from https://github.com/vmware/docker-volume-vsphere/blob/master/tests/constants/dockercli/cmd.go e.g. `TestContainer, vDVSPluginName ` (need to export `vDVSPluginName`)

File: tests/e2e/swarm_test.go [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+
+	status = verification.VerifyDockerService(s.master, s.serviceName, 2)
+	c.Assert(status, Equals, true, Commentf("Service %s is not running", s.serviceName))
+
+	status, host := verification.VerifyDockerContainer(s.dockerHosts, s.serviceName, 2)
+	c.Assert(status, Equals, true, Commentf("Container %s is not running on any hosts", s.serviceName))
+
+	containerName, err = verification.GetContainerName(host, s.serviceName+".1")
+	c.Assert(err, IsNil, Commentf("Failed to retrieve container name: %s", containerName))
+	out, err = dockercli.StopService(host, containerName)
```
Comment (shuklanirdesh82) [inline tests/e2e/swarm_test.go]: nit: new line is required to maintain consistency.

File: tests/utils/verification/volumeproperties.go [hunk 3]
```diff
@@ -97,6 +100,75 @@ func GetVolumePropertiesDockerCli(volName string, hostname string) string {
 	return op
 }
 
+// GetContainerName returns full container name based on given short name
+func GetContainerName(hostName, shortName string) (string, error) {
+	cmd := dockercli.ListContainers + "--filter name='" + shortName + "' --format '{{.Names}}'"
```
Comment (shuklanirdesh82) [inline tests/utils/verification/volumeproperties.go]: I think we should create filter constants for docker volume properties .. e.g. `--format '{{.Names}}` .. it will be easier to make change in one shot to sync with docker framework's change.

File: tests/utils/verification/volumeproperties.go [hunk 4]
```diff
@@ -97,6 +100,75 @@ func GetVolumePropertiesDockerCli(volName string, hostname string) string {
...
+			return status, host
+		}
+		misc.SleepForSec(5)
+	}
+	return false, ""
+}
+
+// allContainersRunning returns true if all replicated containers are up and running; false otherwise.
+func allContainersRunning(dockerHosts []string, serviceName string, replicas int) (bool, string) {
+	//TODO: It looks like all containers will always be spawned on one node. If this assumption
```
Comment (shuklanirdesh82) [inline tests/utils/verification/volumeproperties.go]: It is better to avoid `TODO` in first revision ... do we still need tis TODO?

File: tests/utils/verification/volumeproperties.go [hunk 5]
```diff
@@ -110,6 +182,22 @@ func CheckVolumeAvailability(hostName string, volumeName string) bool {
...
 
+// GetFullVolumeName returns full volume name from the specified VM; return
+// original short name if any error happens
+func GetFullVolumeName(hostName string, volumeName string) string {
+	log.Printf("Fetching full name for volume [%s] from VM [%s]\n", volumeName, hostName)
+
+	cmd := dockercli.ListVolumes + "--filter name='" + volumeName + "' --format '{{.Name}}'"
+	fullName, err := ssh.InvokeCommand(hostName, cmd)
+	if err != nil {
+		log.Println(err)
```
Comment (shuklanirdesh82) [inline tests/utils/verification/volumeproperties.go]: this is redundant .. need to be removed.

File: tests/utils/verification/volumeproperties.go [hunk 6]
```diff
@@ -97,6 +100,75 @@ func GetVolumePropertiesDockerCli(volName string, hostname string) string {
...
+		if status {
+			return status, host
+		}
+		misc.SleepForSec(5)
+	}
+	return false, ""
+}
+
+// allContainersRunning returns true if all replicated containers are up and running; false otherwise.
+func allContainersRunning(dockerHosts []string, serviceName string, replicas int) (bool, string) {
```
Comment (shuklanirdesh82) [inline tests/utils/verification/volumeproperties.go]: `allContainersRunning ` => `areAllContainersRunning ` something more readable

File: tests/utils/verification/volumeproperties.go [hunk 7]
```diff
@@ -97,6 +100,75 @@ func GetVolumePropertiesDockerCli(volName string, hostname string) string {
...
+	//TODO: It looks like all containers will always be spawned on one node. If this assumption
+	//turns out to be false, then the logic below needs to be changed
+	for _, host := range dockerHosts {
+		out, err := GetAllContainers(host, serviceName)
+		if err != nil {
+			log.Println(err)
```
Comment (shuklanirdesh82) [inline tests/utils/verification/volumeproperties.go]: this is redundant ... invokCommand routine takes care of printing error message.

File: tests/utils/verification/volumeproperties.go [hunk 8]
```diff
@@ -97,6 +100,75 @@ func GetVolumePropertiesDockerCli(volName string, hostname string) string {
...
+		if !strings.Contains(out, serviceName+"."+strconv.Itoa(i)) {
+			return false
+		}
+	}
+	return true
+}
+
+// VerifyDockerContainer returns true and the host IP if the containers are running on one of
+// the given docker hosts with specified replicas; otherwise returns false and empty string.
+func VerifyDockerContainer(dockerHosts []string, serviceName string, replicas int) (bool, string) {
```
Comment (shuklanirdesh82) [inline tests/utils/verification/volumeproperties.go]: can you please help me to understand the advantage of returning `host` value here?

File: tests/utils/verification/volumeproperties.go [hunk 9]
```diff
@@ -97,6 +100,75 @@ func GetVolumePropertiesDockerCli(volName string, hostname string) string {
...
+	log.Printf("Checking running containers for docker service [%s] on docker hosts: %v\n", serviceName, dockerHosts)
+
+	//TODO: Need to implement generic polling logic for better reuse
+	for attempt := 0; attempt < maxAttempt; attempt++ {
```
Comment (shuklanirdesh82) [inline tests/utils/verification/volumeproperties.go]: I would suggest to move out this code block into another method and put in place at `misc`.

File: tests/utils/verification/volumeproperties.go [hunk 10]
```diff
@@ -97,6 +100,75 @@ func GetVolumePropertiesDockerCli(volName string, hostname string) string {
...
+
+// GetAllContainers returns all running containers on the give host based on the filtering criteria
+func GetAllContainers(hostName, filter string) (string, error) {
+	cmd := dockercli.ListContainers + "--filter name='" + filter + "' --format '{{.Names}}'"
+	return ssh.InvokeCommand(hostName, cmd)
+}
+
+// VerifyDockerService returns true if the given service is running on
+// the specified VM with specified replicas; false otherwise.
+func VerifyDockerService(hostName, serviceName string, replicas int) bool {
```
Comment (shuklanirdesh82) [inline tests/utils/verification/volumeproperties.go]: as this method is returning `bool` how about renaming to `IsDockerServiceVerified`?

File: tests/utils/verification/volumeproperties.go (hunk 8 above)
Comment (shaominchen) [inline tests/utils/verification/volumeproperties.go]: Please refer to swarm_test line 114 - the returned host is needed for the next steps: query the  container name and stop the service *on this specific host*. My first version of this function does not return host actually. But the test failed in CI setup because the container might be spawned in any hosts. So when verifying the running containers, I need to return the host and then run subsequent steps on this specific host.

Comment (shuklanirdesh82) [inline tests/utils/verification/volumeproperties.go]: got it. Thanks @shaominchen for the explanation.

File: tests/utils/verification/volumeproperties.go [hunk 11]
```diff
@@ -97,6 +100,75 @@ func GetVolumePropertiesDockerCli(volName string, hostname string) string {
...
 }
 
+// GetContainerName returns full container name based on given short name
+func GetContainerName(hostName, shortName string) (string, error) {
+	cmd := dockercli.ListContainers + "--filter name='" + shortName + "' --format '{{.Names}}'"
+	return ssh.InvokeCommand(hostName, cmd)
+}
+
+// GetAllContainers returns all running containers on the give host based on the filtering criteria
+func GetAllContainers(hostName, filter string) (string, error) {
```
Comment (shuklanirdesh82) [inline tests/utils/verification/volumeproperties.go]: I think `/volumeproperties.go` is not the better home to hold such utils. I would suggest to create another file or place into existing suitable place.

File: tests/utils/dockercli/swarm.go [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// This util provides various helper functions for managing docker swarm cluster.
+
+package dockercli
+
+import (
+	"log"
+
```
Comment (pshahzeb) [inline tests/utils/dockercli/swarm.go]: nit: no newline needed

File: tests/utils/dockercli/swarm.go [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+// CreateService creates a docker service
+func CreateService(ip, name, opts string) (string, error) {
+	log.Printf("Creating a docker service [%s] on VM [%s]\n", name, ip)
+	return ssh.InvokeCommand(ip, dockercli.CreateService+"--name "+name+" "+opts)
+}
+
+// ListService lists the tasks of a docker service
+func ListService(ip, name string) (string, error) {
+	log.Printf("Listing docker services running on VM [%s]\n", ip)
```
Comment (pshahzeb) [inline tests/utils/dockercli/swarm.go]: minor: can log the service name as well.

File: tests/utils/verification/volumeproperties.go [hunk 14]
```diff
@@ -97,6 +100,75 @@ func GetVolumePropertiesDockerCli(volName string, hostname string) string {
...
+func allContainersRunning(dockerHosts []string, serviceName string, replicas int) (bool, string) {
+	//TODO: It looks like all containers will always be spawned on one node. If this assumption
+	//turns out to be false, then the logic below needs to be changed
+	for _, host := range dockerHosts {
+		out, err := GetAllContainers(host, serviceName)
+		if err != nil {
+			log.Println(err)
+			continue
+		}
+		if out != "" {
```
Comment (pshahzeb) [inline tests/utils/verification/volumeproperties.go]: as per go suggestions, we can avoid nesting here. 
```
if out == "" {
    continue
}
....
for loop...
```

File: tests/e2e/swarm_test.go [hunk 15]
```diff
@@ -0,0 +1,141 @@
...
+	volumeName  string
+	serviceName string
+}
+
+func (s *SwarmTestSuite) SetUpTest(c *C) {
+	s.esxName = inputparams.GetEsxIP()
+	s.master = inputparams.GetSwarmManager1()
+	s.worker1 = inputparams.GetSwarmWorker1()
+	s.worker2 = inputparams.GetSwarmWorker2()
+	s.dockerHosts = []string{s.master, s.worker1, s.worker2}
```
Comment (pshahzeb) [inline tests/e2e/swarm_test.go]: if you think these vars need to be read just once for the entire test suite, you can have them in SetUpSuite instead of SetupTest.

Comment (shaominchen) [inline tests/e2e/swarm_test.go]: Sure.

File: tests/e2e/swarm_test.go (hunk 1 above)
Comment (shaominchen) [inline tests/e2e/swarm_test.go]: Sure.

File: tests/e2e/swarm_test.go (hunk 2 above)
Comment (shaominchen) [inline tests/e2e/swarm_test.go]: The intention here is to keep the logic of one test step in one single block.

File: tests/utils/dockercli/swarm.go (hunk 12 above)
Comment (shaominchen) [inline tests/utils/dockercli/swarm.go]: Not sure why gofmt does not take care of this. Fixed.

File: tests/utils/dockercli/swarm.go (hunk 13 above)
Comment (shaominchen) [inline tests/utils/dockercli/swarm.go]: Sure.

File: tests/utils/verification/volumeproperties.go (hunk 3 above)
Comment (shaominchen) [inline tests/utils/verification/volumeproperties.go]: I didn't define constant for "--format" because we are using quite some different formats: .Name, .Names, .Status, etc. As for "--filter", it needs a input parameter - defining a constant like "--filter name='" looks weird, as we need attach the matching quote ' separately.

File: tests/utils/verification/volumeproperties.go (hunk 11 above)
Comment (shaominchen) [inline tests/utils/verification/volumeproperties.go]: Make sense. Will add a separate dockercontainer.go file.

File: tests/utils/verification/volumeproperties.go (hunk 10 above)
Comment (shaominchen) [inline tests/utils/verification/volumeproperties.go]: Renamed it to IsDockerServiceRunning.

File: tests/utils/verification/volumeproperties.go (hunk 9 above)
Comment (shaominchen) [inline tests/utils/verification/volumeproperties.go]: We should refactor this (and existing VerifyDetachedStatus) in a separate change.

File: tests/utils/verification/volumeproperties.go (hunk 6 above)
Comment (shaominchen) [inline tests/utils/verification/volumeproperties.go]: Sure - renamed it to isContainerRunning.

File: tests/utils/verification/volumeproperties.go (hunk 4 above)
Comment (shaominchen) [inline tests/utils/verification/volumeproperties.go]: OK, removed it for now.

File: tests/utils/verification/volumeproperties.go (hunk 7 above)
Comment (shaominchen) [inline tests/utils/verification/volumeproperties.go]: Removed.

File: tests/utils/verification/volumeproperties.go (hunk 14 above)
Comment (shaominchen) [inline tests/utils/verification/volumeproperties.go]: Fixed.

File: tests/utils/verification/volumeproperties.go (hunk 5 above)
Comment (shaominchen) [inline tests/utils/verification/volumeproperties.go]: Removed.

File: tests/e2e/swarm_test.go (hunk 2 above)
Comment (govint) [inline tests/e2e/swarm_test.go]: Please see issue #1315, worth adding a test case that verifies vsphere volumes as "external" in the compose file and verify that the volume on a particular datastore is attached to the VM/container.

File: tests/utils/verification/volumeproperties.go [hunk 16]
```diff
@@ -29,10 +29,6 @@ import (
 	"github.com/vmware/docker-volume-vsphere/tests/utils/ssh"
 )
 
-const (
```
Comment (shuklanirdesh82) [inline tests/utils/verification/volumeproperties.go]: It is better to have constant to modify the attempt count. I would suggest to add this back; it will be easier to adjust in the future instead of skimming line of code.

File: tests/utils/verification/dockercontainer.go [hunk 17]
```diff
@@ -0,0 +1,95 @@
...
+	"strconv"
+
+	"github.com/vmware/docker-volume-vsphere/tests/constants/dockercli"
+	util "github.com/vmware/docker-volume-vsphere/tests/utils/dockercli"
+	"github.com/vmware/docker-volume-vsphere/tests/utils/misc"
+	"github.com/vmware/docker-volume-vsphere/tests/utils/ssh"
+)
+
+// GetContainerName returns full container name based on given short name
+func GetContainerName(hostName, shortName string) (string, error) {
```
Comment (shuklanirdesh82) [inline tests/utils/verification/dockercontainer.go]: `GetContainerName ` & `GetAllContainers` should move out from verifier package ... both of them spoils the context of `verifier` .. /util/dockercli would be a better home.

File: tests/utils/verification/volumeproperties.go (hunk 16 above)
Comment (shaominchen) [inline tests/utils/verification/volumeproperties.go]: I removed this because there's a conflict on this constant name after splitting this file into two. Also we need to refactor this part anyway (I've already added TODO on both functions). We should have a generic poller function, which accepts a maximum time limit instead of the maximum attempts as the input.

File: tests/utils/verification/dockercontainer.go (hunk 17 above)
Comment (shaominchen) [inline tests/utils/verification/dockercontainer.go]: Good idea! Fixed.

File: tests/e2e/swarm_test.go (hunk 15 above)
Comment (shaominchen) [inline tests/e2e/swarm_test.go]: I will add a GetSwarmNodes function in inputparams. I don't follow why we need a map here.

File: tests/e2e/swarm_test.go [hunk 18]
```diff
@@ -0,0 +1,142 @@
...
+
+var _ = Suite(&SwarmTestSuite{})
+
+// Test vDVS usage in swarm mode
+//
+// Test steps:
+// 1. Create a docker service with replicas setting to 1
+// 2. Verify the service is up and running with one node
+// 3. Verify one container is spawned
+// 5. Verify the volume is in attached status
```
Comment (lipingxue) [inline tests/e2e/swarm_test.go]: The --replica option in swarm is not used to scale a stateful application. Instead, it should used to scale a stateless application. So why do we try to add the test with --replica 2 here? Any specific reason? @ashahi1

I think we should just verify the fail-over case here.

File: tests/e2e/swarm_test.go (hunk 2 above)
Comment (shaominchen) [inline tests/e2e/swarm_test.go]: First of all, if we need any test to cover new scenarios, that should be done in a separate PR. Moreover, I still don't follow why we need a e2e test for #1315. We can discuss that in the issue itself instead of in this PR.

File: tests/utils/dockercli/swarm.go [hunk 19]
```diff
@@ -0,0 +1,68 @@
+// Copyright 2017 VMware, Inc. All Rights Reserved.
+//
```
Comment (shaominchen) [inline tests/utils/dockercli/swarm.go]: As we discussed offline, we should keep each PR small enough so that it’s easier to review and also keep the dependency to the minimum. The current PR is probably a special case as there's only one test case, but it's still pretty big and complicated. Since the review is almost done, I'd prefer to keep the current PR as is instead of wasting time to split it into two.

**Comment by shuklanirdesh82** [comment]
@shaominchen If this PR is targeted for #1249 ... it is better to add ```fixes #1249``` to close the issue automatically with merge.

File: tests/utils/verification/dockercontainer.go [hunk 20]
```diff
@@ -0,0 +1,95 @@
...
+		if status {
+			return status, host
+		}
+		misc.SleepForSec(5)
+	}
+	return false, ""
+}
+
+// isContainerRunning returns true if all replicated containers are up and running; false otherwise.
+func isContainerRunning(dockerHosts []string, serviceName string, replicas int) (bool, string) {
```
Comment (lipingxue) [inline tests/utils/verification/dockercontainer.go]: Nit：function name change to "IsContainerRunning", which is consistent with other func names

File: tests/e2e/swarm_test.go (hunk 18 above)
Comment (shaominchen) [inline tests/e2e/swarm_test.go]: @lipingxue Can we add this comment in the issue #601 itself? We can discuss your comment in the issue and once we have an agreement we can come back to update this test.

File: tests/utils/verification/dockercontainer.go (hunk 20 above)
Comment (shaominchen) [inline tests/utils/verification/dockercontainer.go]: This is a convention in Go language: unexported identifiers should start with small letter; exported identifiers should start with capital letter.

File: tests/e2e/swarm_test.go (hunk 18 above)
Comment (lipingxue) [inline tests/e2e/swarm_test.go]: OK. I reopened issue #601 to add my comments.

File: tests/utils/verification/dockercontainer.go (hunk 20 above)
Comment (lipingxue) [inline tests/utils/verification/dockercontainer.go]: ok

**Comment by shuklanirdesh82** [comment]
@shaominchen  Please revisit the commit history and clean that up correctly.

File: vmdk_plugin/Makefile [hunk 21]
```diff
@@ -287,7 +287,7 @@ CLEANESX_SH   := $(DEPLOY_TOOLS) cleanesx
 deploy-esx: clean-esx
 	$(DEPLOY_ESX_SH) "$(ESX)" "$(VIB_BIN)"
 
-VMS= $(VM1) $(VM2) $(WORKER2)
+VMS= $(VM1) $(VM2)
```
Comment (shuklanirdesh82) [inline vmdk_plugin/Makefile]: this is required, isn't it?

Comment (shaominchen) [inline vmdk_plugin/Makefile]: Yes, will fix.
